feat(flow): implement adaptive throttler

kxrxh opened the pull request:
## Motivation
* Implements #191

## Modifications
* **AdaptiveThrottler**: Added a new flow component that regulates message processing speed using a feedback loop.
* **Resource Monitoring**: Implemented `ResourceMonitor` with pluggable `cpuUsageSampler` (heuristic & measured modes) and `memoryReader` (cgroup v1/v2 & host support).
* **Configuration**: Introduced `AdaptiveThrottlerConfig` for tuning thresholds, adaptation steps, and sampling intervals.
* **Testing**: Added unit tests for the throttler logic and resource samplers to ensure correct behavior under load.

## Verify change
* [x] Make sure the change passes the CI checks.
* [x] Do tests in Docker, K8S, host environments
* [x] Test functionality using demo application

reugn committed:
fix(flow): emit partial sliding window on early close (#192)

kxrxh committed:
feat(flow): implement adaptive throttler with resource monitoring and throughput control

kxrxh committed:
docs(README): add AdaptiveThrottler description to documentation

kxrxh committed:
fix: resolve all golangci-lint issues

- Fix gocritic issues (elseif, ifElseChain)
- Fix gosec G115 integer overflow
- Fix revive unused parameter issues
- Fix unconvert unnecessary conversions
- Fix line length violations (>120 chars)
- Rename CPUUsageModeRusage → CPUUsageModeReal
- Move goroutineHeuristicSampler to dedicated file
- Update gopsutil from v4.25.10 → v4.24.6 for Go 1.21 compatibility

kxrxh committed:
refactor(adaptive_throttler): streamline main function by removing commented-out configuration examples

kxrxh committed:
refactor: rename CPUUsageModeReal to CPUUsageModeMeasured

- Rename CPUUsageModeReal to CPUUsageModeMeasured for clarity
- Add clarifying comment to setSystemMemoryReader function

The new name better reflects that even gopsutil measurements are
approximations rather than absolute 'real' values.

kxrxh committed:
feat(adaptive_throttler): enhance resource monitoring for containerized environments

- Add cgroup v2/v1 auto-detection for accurate memory stats in containers
- Normalize CPU usage sampling by core count for better accuracy
- Support custom memory readers in ResourceMonitor for flexible deployments
- Add comprehensive system_memory tests with mock file systems
- Improve test reliability with polling-based waiting instead of fixed sleeps

kxrxh committed:
Merge pull request #1 from reugn/master

fix(flow): emit partial sliding window on early close (#192)

kxrxh committed:
feat(adaptive_throttler): add demo for adaptive throttling pipeline

- Implement a demo application showcasing the adaptive throttler in action

reugn commented on README.md:
```diff
@@ -43,6 +43,7 @@ processing operations. These building blocks can be used to transform and manipu
 - **Flatten<sup>1</sup>:** Flattens a stream of slices of elements into a stream of elements.
 - **Batch:** Breaks a stream of elements into batches based on size or timing.
 - **Throttler:** Limits the rate at which elements are processed.
+- **AdaptiveThrottler:** Limits the rate at which elements are processed based on the current system resource utilization (CPU and memory usage)
```
```suggestion
- **AdaptiveThrottler:** Limits the rate at which elements are processed based on the current system resource utilization (CPU and memory usage).
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+}
+
+// resourceMonitorInterface defines the interface for resource monitoring
+type resourceMonitorInterface interface {
```
```suggestion
type resourceMonitor interface {
```
To keep it concise.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+type resourceMonitorInterface interface {
+	GetStats() ResourceStats
+	IsResourceConstrained() bool
+	Close()
```
Add documentation to the public methods.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+}
+
+// Validate checks that the configuration is valid and returns an error if not
+func (c *AdaptiveThrottlerConfig) Validate() error {
```
If it is planned to be called only in the constructor, it is better to unexport the function.

kxrxh replied:
I have to leave it that way, because otherwise golangci-lint throws an error.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+		return fmt.Errorf("invalid throughput bounds: min=%d, max=%d", c.MinThroughput, c.MaxThroughput)
+	}
+	if c.AdaptationFactor <= 0 || c.AdaptationFactor >= 1 {
+		return fmt.Errorf("AdaptationFactor must be in (0, 1), got %f", c.AdaptationFactor)
```
```suggestion
		return fmt.Errorf("invalid AdaptationFactor: %f, must be in (0, 1)", c.AdaptationFactor)
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+var _ streams.Flow = (*AdaptiveThrottler)(nil)
+
+// NewAdaptiveThrottler creates a new adaptive throttler
+func NewAdaptiveThrottler(config AdaptiveThrottlerConfig) *AdaptiveThrottler {
```
Consider taking a pointer as the parameter and using the default configuration if nil.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+// NewAdaptiveThrottler creates a new adaptive throttler
+func NewAdaptiveThrottler(config AdaptiveThrottlerConfig) *AdaptiveThrottler {
+	if err := config.Validate(); err != nil {
+		panic(fmt.Sprintf("invalid config: %v", err))
```
Document all the panics.

kxrxh replied:
> Document all the panics.

However, do I need to use panics here? Or I can instead return an error and let the user handle it (panics now done based on your throttle implementation)

reugn replied:
Not returning errors enables the seamless inlining of new flows into the pipeline.
Given the wide range of possible errors, returning the error would be better in this particular case.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+	currentRate := float64(at.currentRate.Load())
+	targetRate := currentRate
+
+	if constrained {
```
Document the constrained flag.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+// adaptRate adjusts the throughput rate based on current resource usage
+func (at *AdaptiveThrottler) adaptRate() {
+	at.adaptMu.Lock()
+	defer at.adaptMu.Unlock()
```
Since `adaptRate` cannot be called concurrently and this mutex is only used here, what is its purpose?

kxrxh replied:
> Since `adaptRate` cannot be called concurrently and this mutex is only used here, what is its purpose?

The adaptMu mutex in adaptRate() protects against race conditions with shared state.
Without it, quotaExceeded() (called from concurrent emit() goroutines) could read inconsistent values - seeing new maxElements with old counter.

reugn replied:
There are no race conditions here, and this mutex can be safely removed.
- `adaptRate` is only called from `adaptRateLoop` (without concurrent calls) (line 205)
- `adaptRateLoop` runs in a single goroutine (line 186)

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+		if !at.quotaExceeded() {
+			at.counter.Add(1)
+			at.out <- element
+			return
```
Should be `continue`?

kxrxh replied:
> Should be `continue`?

The function waits to send one element, but only when there is available "quota".
If quota exceeded -> blocks and waits for quota reset signal.
If quota available -> sends immediately and exits.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,413 @@
+	}
+
+	at.stop()
+}
```
Are those needed for tests only?

kxrxh replied:
> Are those needed for tests only?

Yes, GetCurrentRate() and GetResourceStats() are only used in the demo.go, so they can be deleted.

reugn commented on flow/cpu_sampler.go:
```diff
@@ -0,0 +1,74 @@
+package flow
```
Move all helper structs under the new `flow/internal/adaptive`  package to hide their implementation details from the public API.

reugn commented on flow/cpu_sampler_heuristic.go:
```diff
@@ -0,0 +1,57 @@
+// goroutineHeuristicSampler uses goroutine count as a CPU usage proxy
+type goroutineHeuristicSampler struct{}
+
+func (s *goroutineHeuristicSampler) Sample(_ time.Duration) float64 {
```
Use the value accessor.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,305 @@
+	memoryReader func() (float64, error),
+) *ResourceMonitor {
+	if sampleInterval <= 0 {
+		panic(fmt.Sprintf("invalid sampleInterval: %v", sampleInterval))
```
Document all the panics.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,305 @@
+
+	rm.initSampler()
+
+	rm.stats.Store(rm.collectStats())
```
Can be moved into `initSampler`.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,305 @@
+		if percent > 100 {
+			return 100
+		}
+		return percent
```
Consider extracting to a utility function and reusing.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,305 @@
+	case stats.CPUUsagePercent < 0:
+		stats.CPUUsagePercent = 0
+	case stats.CPUUsagePercent > 100:
+		stats.CPUUsagePercent = 100
```
Use the validate util function.

reugn commented on flow/system_memory.go:
```diff
@@ -0,0 +1,219 @@
+	"strings"
+	"sync/atomic"
+
+	"github.com/shirou/gopsutil/v4/mem"
```
The main module is kept zero-dependency. This library is used as a fallback and can be replaced with standard-library packages. Add platform-specific implementations if needed.

kxrxh replied:
> The main module is kept zero-dependency. This library is used as a fallback and can be replaced with standard-library packages. Add platform-specific implementations if needed.

I am currently rewriting this module without dependencies. To implement this for darwin platform, I need a library "golang.org/x/sys/unix ". Can I include it in the root go.mod file?

reugn replied:
Prioritize using core packages, such as `os` . If these are insufficient, return a memory usage percentage of 0, signifying no concern.

kxrxh committed:
refactor(adaptive_throttler): update config validation message

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
docs(README): fix punctuation in AdaptiveThrottler description

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
refactor(adaptive_throttler): rename resourceMonitorInterface to resourceMonitor

kxrxh committed:
refactor(resource_monitor): move stats collection to initSampler

reugn commented on flow/system_memory_test.go:
```diff
@@ -0,0 +1,404 @@
+
+// Test concurrent access to system memory reader
+func TestConcurrentSystemMemoryAccess(t *testing.T) {
+	done := make(chan bool, 10)
```
Consider using `sync.WaitGroup` instead.

reugn commented on flow/system_memory.go:
```diff
@@ -0,0 +1,219 @@
+func init() {
+	systemMemoryProvider.Store(systemMemoryReader(readSystemMemoryAuto))
+	fileSystemProvider.Store(FileSystem(osFileSystem{}))
+}
```
Look for alternatives for this `init` call. This functionality may not be in use at all.

kxrxh committed:
refactor(adaptive_throttler): extract system monitoring into internal/sysmonitor package

Refactor CPU and memory monitoring code from the flow package into a dedicated internal/sysmonitor package. This improves code organization, testability, and platform-specific implementation support.

Key changes:
- Move CPU sampling logic to internal/sysmonitor with platform-specific implementations for Darwin, Linux, Windows, and fallback
- Move system memory monitoring to internal/sysmonitor with support for:
  - Linux: /proc/meminfo, cgroup v1, and cgroup v2 detection
  - Darwin: Mach kernel APIs with CGO
  - Windows: GlobalMemoryStatusEx API
  - Fallback: Error handling for unsupported platforms
  - Move heuristic CPU sampler from flow/cpu_sampler_heuristic.go to internal/sysmonitor/heuristic.go
- Update adaptive_throttler and resource_monitor to use new package
- Remove flow/cpu_sampler.go, flow/system_memory.go

kxrxh committed:
refactor(sysmonitor): improve CPU time retrieval for current process

Updated the getProcessCPUTimes function to use GetCurrentProcess for the current process, enhancing reliability and eliminating the need to open a process handle. This change simplifies the code and improves performance when retrieving CPU times.

kxrxh committed:
feat(adaptive_throttler): enhance demo with CPU workload simulation

kxrxh committed:
docs(adaptive_throttler): add clarification comment for constrained variable

kxrxh commented:
@reugn Changes applied. Ready for re-review.

reugn commented on internal/sysmonitor/cpu_linux.go:
```diff
@@ -0,0 +1,191 @@
+	wallTimeSeconds := elapsed.Seconds()
+
+	// Normalized to 0-100% (divides by numCPU for system-wide metric)
+	numcpu := runtime.NumCPU()
```
Add:
```go
if numcpu <= 0 {
    numcpu = 1 // Safety check
}
```

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+	}
+
+	// Parse memory.stat to find reclaimable memory (inactive_file)
+	inactiveFile, _ := readCgroupStatWithFS(fs, "/sys/fs/cgroup/memory.stat", "inactive_file")
```
Unhandled error.

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+			}
+		}
+	}
+	return 0, fmt.Errorf("key %q not found in %s", key, path)
```
Check for scanner errors before returning:
```go
if err := scanner.Err(); err != nil {
	return 0, fmt.Errorf("error reading %s: %w", path, err)
}
```

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+// GetSystemMemory returns the current system memory statistics.
+// This function auto-detects the environment (cgroup v2, v1, or host)
+// and returns appropriate memory information.
+func GetSystemMemory() (SystemMemory, error) {
```
Consider simplifying this double-lock pattern.

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+	return readCgroupV1MemoryWithFS(fileSystem)
+}
+
+func readCgroupV1MemoryWithFS(fs FileSystem) (SystemMemory, error) {
```
This function is almost identical to `readCgroupV2MemoryWithFS` and can be parameterized to reduce code duplication.

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+
+	// Check for "unlimited" (random huge number in V1)
+	if limit > (1 << 60) {
+		return SystemMemory{}, os.ErrNotExist
```
`os.ErrNotExist` seems inappropriate in this context.

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+func readCgroupValueWithFS(fs FileSystem, path string) (uint64, error) {
+	data, err := fs.ReadFile(path)
+	if err != nil {
+		return 0, err
```
Improve error handling by wrapping the error with additional context (across the package):
```go
fmt.Errorf("failed to read file %s: %w", path, err)
```

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+		}
+
+		// Convert from kB to bytes
+		value *= 1024
```
Could overflow if `value > (2^64 - 1) / 1024`.

reugn commented on internal/sysmonitor/memory_linux.go:
```diff
@@ -0,0 +1,266 @@
+			found)
+	}
+
+	return SystemMemory{
```
Verify that available doesn't exceed total.

reugn commented on internal/sysmonitor/memory_windows.go:
```diff
@@ -0,0 +1,70 @@
+	ret, _, err := procGlobalMemoryStatusEx.Call(uintptr(unsafe.Pointer(&memStatus)))
+
+	// If the function fails, the return value is zero.
+	if ret == 0 {
```
```suggestion
	if ret == 0 || err != nil {
```

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,265 @@
+		panic(fmt.Sprintf("invalid cpuThreshold: %f, must be between 0 and 100", cpuThreshold))
+	}
+
+	rm := &ResourceMonitor{
```
It would be beneficial to allow a shared singleton `ResourceMonitor` between multiple instances of `AdaptiveThrottler` for consistent metrics and less overhead from multiple samplers.

Need to consider possible different thresholds and when to close the shared monitor.

kxrxh replied:
Implemented shared ResourceMonitor for all AdaptiveThrottler instances. Now they all use the same monitor instead of each creating their own

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+// DefaultAdaptiveThrottlerConfig returns sensible defaults for most use cases
+func DefaultAdaptiveThrottlerConfig() AdaptiveThrottlerConfig {
```
Return a pointer, as the constructor expects a pointer parameter.

reugn commented on go.sum:
Empty `go.sum` can be removed.

reugn commented on flow/util.go:
```diff
@@ -177,3 +178,23 @@ func Flatten[T any](parallelism int) streams.Flow {
+		return 0
+	}
+	return clampPercent(percent)
+}
```
Move those under `resource_monitor.go` where they belong. I will consider renaming this file from `util` to `operators` or similar.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,265 @@
+
+	rm.initSampler()
+
+	go rm.monitor()
```
```suggestion
    // start periodically collecting resource statistics
	go rm.monitor()
```

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,265 @@
+	cpuMode         CPUUsageMode
+
+	// Current stats (atomic for thread-safe reads)
+	stats atomic.Value // *ResourceStats
```
Use the idiomatic `atomic.Pointer` type:
```go
stats atomic.Pointer[ResourceStats]
```
for type safety and better performance.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+// adaptRate adjusts the throughput rate based on current resource usage
+func (at *AdaptiveThrottler) adaptRate() {
```
Explore options to simplify and improve the readability of this method.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+		MinThroughput:       10,                     // Reasonable minimum throughput
+		MaxThroughput:       500,                    // More conservative maximum
+		SampleInterval:      200 * time.Millisecond, // Less frequent sampling
+		BufferSize:          500,                    // Match max throughput for 1 second buffer at max rate
```
Add units to the field name or document.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+	if config.MaxCPUPercent <= 0 || config.MaxCPUPercent > 100 {
+		return nil, fmt.Errorf("invalid MaxCPUPercent: %f", config.MaxCPUPercent)
+	}
+	if config.BufferSize < 1 {
```
Introduce a `maxBufferSize` and validate to avoid unbounded memory allocation.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+	// Default: 0.3 (max 30% change per cycle)
+	MaxRateChangeFactor float64
+
+	// MemoryReader provides memory usage percentage for containerized deployments.
```
```suggestion
	// MemoryReader is a user-provided custom function that returns memory usage percentage.
    // This can be particularly useful for containerized deployments or other environments
    // where standard system memory readings may not accurately reflect container-specific
    // usage.
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+	// CPU usage sampling mode.
+	//
+	// CPUUsageModeHeuristic: Estimates CPU usage using a simple heuristic (goroutine count), suitable for platforms
+	// where accurate process CPU measurement is not supported.
```
```suggestion
	// CPUUsageModeHeuristic: Estimates CPU usage using a simple heuristic (goroutine count),
	// suitable for platforms where accurate process CPU measurement is not supported.
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+
+	// MemoryReader provides memory usage percentage for containerized deployments.
+	// If nil, system memory will be read via mem.VirtualMemory().
+	// Should return memory used percentage (0-100).
```
```suggestion
	// Must return memory used percentage (0-100).
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+		maxReduction := currentRate * at.config.MaxRateChangeFactor
+		reduction := math.Min(currentRate*reductionFactor, maxReduction)
+
+		targetRate = currentRate - reduction
```
Check for negative value.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,415 @@
+	if config.BufferSize < 1 {
+		return nil, fmt.Errorf("invalid BufferSize: %d", config.BufferSize)
+	}
+	if config.SampleInterval <= 0 {
```
Consider adding `minSampleInterval` validation.

reugn commented on examples/adaptive_throttler/demo/demo.go:
```diff
@@ -0,0 +1,147 @@
+	"github.com/reugn/go-streams/flow"
+)
+
+func main() {
```
```suggestion
// main demonstrates adaptive throttling with simulated resource pressure.
// The demo:
// 1. Produces 250 elements in bursts
// 2. Processes elements with CPU-intensive work (50ms each)
// 3. Simulates increasing memory pressure as more elements are processed
// 4. The adaptive throttler adjusts throughput based on CPU/memory usage
// 5. Stats are logged every 500ms showing rate adaptation
func main() {
```

reugn commented on examples/go.mod:
```diff
@@ -80,6 +80,7 @@ require (
 	github.com/go-jose/go-jose/v4 v4.0.4 // indirect
 	github.com/go-logr/logr v1.4.2 // indirect
 	github.com/go-logr/stdr v1.2.2 // indirect
+	github.com/go-ole/go-ole v1.3.0 // indirect
```
All added indirect dependencies can be reverted.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,265 @@
+			available = sysStats.Total
+		}
+
+		// Defensive programming: avoid division by zero
```
```suggestion
		// avoid division by zero
```

reugn commented on flow/resource_monitor_test.go:
```diff
@@ -0,0 +1,299 @@
+}
+
+func TestResourceMonitor_UsesSystemMemoryStats(t *testing.T) {
+	t.Helper()
```
This is not a helper function.

reugn commented on flow/adaptive_throttler.go:
Improve test coverage.

reugn commented on flow/resource_monitor.go:
Improve test coverage.

kxrxh committed:
fix(sysmonitor): add safety checks for CPU count and handle memory stat errors

- Implemented a safety check to ensure the number of CPUs is at least 1 in both Darwin and Linux CPU sampling functions.
- Updated memory reading functions to handle potential errors gracefully, defaulting to 0 for unavailable stats.

kxrxh committed:
refactor(sysmonitor): unify memory reading functions for cgroup v1 and v2

kxrxh committed:
refactor(sysmonitor): streamline memory reading logic and remove unused cpuTimesStat struct

kxrxh committed:
refactor(sysmonitor): additional context for errors was added

kxrxh committed:
refactor(sysmonitor): remove outdated validation comment in getProcessCPUTimes function

kxrxh committed:
fix(sysmonitor): add overflow check for memory value conversion

kxrxh committed:
fix(sysmonitor): enhance error handling in memory status retrieval

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
docs(adaptive_throttler): update demo comments to clarify adaptive throttling process

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
refactor(resource_monitor_test): remove unnecessary helper call in test function

kxrxh committed:
Merge branch 'feature/adaptive-throttler' of https://github.com/kxrxh/go-streams into feature/adaptive-throttler

kxrxh committed:
refactor(resource_monitor): simplify comment for division by zero check

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
chore(dependencies): downgrade Go version to 1.23.0 and update indirect dependencies

kxrxh committed:
feat(adaptive_throttler): enforce minimum sample interval to reduce CPU overhead

kxrxh committed:
fix(adaptive_throttler): prevent negative target rate in rate adaptation

kxrxh committed:
docs(adaptive_throttler): clarify comment on MemoryReader function requirement

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
docs(adaptive_throttler): format comment for CPU usage sampling modes

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
feat(resource_monitor): add comment to indicate start of periodic resource statistics collection

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
docs(adaptive_throttler): update MemoryReader comment for clarity on its functionality

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
chore(dependencies): remove go.sum and update go.mod

kxrxh committed:
refactor(resource_monitor): move clampPercent and validatePercent functions to resource_monitor.go

kxrxh committed:
refactor(flow): rename util.go to operators.go for better clarity

kxrxh committed:
refactor(adaptive_throttler): remove unnecessary pointer dereference in NewAdaptiveThrottler calls

kxrxh committed:
refactor(resource_monitor): change stats from atomic.Value to atomic.Pointer for improved type safety

kxrxh committed:
refactor(adaptive_throttler): remove unnecessary mutex for rate adaptation

kxrxh committed:
refactor(adaptive_throttler): reorder functions and improved adaptRate readability

kxrxh committed:
feat(adaptive_throttler): add MaxBufferSize configuration and validation checks

kxrxh committed:
refactor(sysmonitor): simplify memory initialization by removing unnecessary mutex locking

kxrxh committed:
fix(sysmonitor): ensure available memory does not exceed total memory

kxrxh committed:
refactor(adaptive_throttler): fix golangci-lint errors

kxrxh committed:
refactor(adaptive_throttler): change validateConfig function to a method on AdaptiveThrottlerConfig

kxrxh committed:
refactor(adaptive_throttler, resource_monitor): enhance configuration documentation and improve code clarity

kxrxh committed:
test: add comprehensive test coverage for adaptive throttler and system monitor

kxrxh committed:
test(resource_monitor): add tests for clamping and validation of percentage values

kxrxh committed:
test(sysmonitor): refactor memory tests to use a helper function to fix dupl error

kxrxh committed:
refactor(adaptive_throttler, resource_monitor): implement shared monitor registry with reference counting

- Replace singleton ResourceMonitor with shared registry pattern for efficient resource sharing
- Add dynamic interval updates and improved CPU mode switching
- Enhance memory sampling with better fallback handling
- Add InDelta assertion helper for precise float comparisons
- Comprehensive test refactoring with improved mocking and coverage
- Update adaptive throttler tests to use new resource monitor architecture

kxrxh commented:
@reugn  Can you please re-run workflows (codecov, golangci-lint)?

kxrxh committed:
refactor(sysmonitor): remove SetMemoryReader function from memory platform files

kxrxh committed:
refactor: enhance system monitoring and adaptive throttling

- Restructure sysmonitor package with platform-specific CPU/memory monitoring
- Rename heuristic.go to cpu_heuristic.go for clarity
- Add comprehensive test coverage for CPU and memory monitoring across platforms
- Enhance ResourceMonitor with registry pattern for shared monitoring
- Improve AdaptiveThrottler with hysteresis, recovery thresholds, and smoothing
- Add test helpers and mock filesystem for better testing
- Update adaptive throttler example with improved configuration

kxrxh committed:
test(sysmonitor): add comprehensive tests for Windows CPU and memory monitoring

kxrxh committed:
fix(adaptive_throttler): make close public and ensure proper resource cleanup

kxrxh committed:
fix(adaptive_throttler, resource_monitor): correct loop iteration and enhance resource monitoring setup

kxrxh committed:
test: refactor adaptive throttler and monitor suites

kxrxh committed:
fix: stop resource monitor deterministically

kxrxh committed:
refactor: wait for monitor shutdown with waitgroup

kxrxh committed:
refactor(resource_monitor): simplify stop timer handling during resource acquisition

kxrxh committed:
refactor(adaptive_throttler): share rate calc, allow CPU limit 0, guard missing stats, add tests

kxrxh commented:
@reugn Can you please check once again?

reugn commented on flow/operators.go:
Revert renaming the file.

reugn commented on internal/testutil/sysmonitor.go:
Consider extracting to a `testutil` internal package.

reugn commented on flow/adaptive_throttler.go:
See the [uncovered code lines](https://app.codecov.io/gh/reugn/go-streams/blob/kxrxh%3Afeature%2Fadaptive-throttler/flow%2Fadaptive_throttler.go). There are some that can be easily covered.

reugn commented on flow/resource_monitor.go:
See the [uncovered code lines](https://app.codecov.io/gh/reugn/go-streams/blob/kxrxh%3Afeature%2Fadaptive-throttler/flow%2Fresource_monitor.go).

kxrxh replied:
Please, rerun Codecov

reugn commented on flow/adaptive_throttler_test_helpers.go:
Extract to a `testutil` internal package.

reugn commented on flow/resource_monitor_test_helpers.go:
Extract to a `testutil` internal package.

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,377 @@
+		config = DefaultAdaptiveThrottlerConfig()
+	}
+	if err := config.validate(); err != nil {
+		return nil, err
```
```suggestion
		return nil, fmt.Errorf("invalid adaptive throttler config: %w", err)
```

reugn commented on flow/adaptive_throttler.go:
```diff
@@ -0,0 +1,377 @@
+	}
+}
+
+func (at *AdaptiveThrottler) Close() {
```
Doesn't close the `at.in` channel to unblock `pipelineLoop`. Consider making this method private, as it is used only for testing, and documenting its use.

reugn commented on flow/resource_monitor.go:
```diff
@@ -0,0 +1,419 @@
+	switch {
+	// Check if a custom memory reader is provided
+	case rm.memoryReader != nil:
+		if mem, err := rm.memoryReader(); err == nil {
```
Consider adding panic recovery for the custom `memoryReader` call.

reugn commented on internal/sysmonitor/cpu_linux.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+// Sample returns the CPU usage percentage since the last sample
+func (s *linuxProcessSampler) Sample(deltaTime time.Duration) float64 {
```
There is a possible race condition if `Sample` is called concurrently (same in other samplers). If this is by design, document as not thread-safe.

kxrxh replied:
Yes, it is not thread-safe. In our main usage (ResourceMonitor), Sample is invoked from a single monitoring goroutine. So, I documented the interface as not safe for concurrent use (callers should wrap with an external mutex if needed).

reugn commented on internal/sysmonitor/memory_windows.go:
```diff
@@ -0,0 +1,54 @@
+	ret, _, err := procGlobalMemoryStatusEx.Call(uintptr(unsafe.Pointer(&memStatus)))
+
+	// If the function fails, the return value is zero.
+	if ret == 0 {
```
```suggestion
	if err != nil || ret == 0 {
```

kxrxh replied:
On Windows, this should not be checked with err != nil: Go docs say “the returned error is always non-nil … callers must inspect the primary return value … before consulting the error”

[Doc ref](https://go.dev/src/syscall/dll_windows.go#:~:text=%2F%2F%20The%20returned%20error%20is%20always%20non%2Dnil%2C%20constructed%20from%20the%20result%20of%20GetLastError%2E%20153&text=%2F%2F%20Callers%20must%20inspect%20the%20primary%20return%20value%20to%20decide%20whether%20an%20error%20occurred%20154&text=%2F%2F%20%28according%20to%20the%20semantics%20of%20the%20specific%20function%20being%20called%29%20before%20consulting%20155&text=%2F%2F%20the%20error%2E%20The%20error%20always%20has%20type%20%5BErrno%5D%2E%20156&text=%2F%2F,-157)

reugn commented:
Hi @kxrxh, please let me know if I can help with any of the requested changes.

kxrxh committed:
refactor(flow): revert file renaming

kxrxh committed:
docs(cpu): add thread-safety notes to CPU usage documentation

kxrxh committed:
test(adaptive_throttler): enhance validation tests and moved mocks/helpers to testutil package, where possible

kxrxh committed:
refactor(adaptive_throttler): update config validation message

Co-authored-by: Eugene R. <reugpro@gmail.com>

kxrxh committed:
fix(flow): make AdaptiveThrottler close() private and fix channel cleanup

kxrxh committed:
fix(resource_monitor): handle panic in custom memory reader

reugn commented:
@kxrxh, thank you for this contribution!
